src/ReasonReactRouter.re

```diff
@@ -0,0 +1,170 @@
+    };
+  tolist(Array.length(a) - 1, []);
+};
+/* if we ever roll our own parser in the future, make sure you test all url combinations
+   e.g. foo.com/?#bar
+   */
+/* sigh URLSearchParams doesn't work on IE11, edge16, etc. */
+/* actually you know what, not gonna provide search for now. It's a mess.
+   We'll let users roll their own solution/data structure for now */
+let path = () =>
+  switch ([%external window]) {
+  | None => []
+  | Some((window: Dom.window)) =>
+    switch (window |> location |> pathname) {
+    | ""
+    | "/" => []
+    | raw =>
+      /* remove the preceeding /, which every pathname seems to have */
+      let raw = Js.String.sliceToEnd(~from=1, raw);
+      /* remove the trailing /, which some pathnames might have. Ugh */
+      let raw =
+        switch (Js.String.get(raw, Js.String.length(raw) - 1)) {
+        | "/" => Js.String.slice(~from=0, ~to_=-1, raw)
+        | _ => raw
+        };
+      raw |> Js.String.split("/") |> arrayToList;
+    }
+  };
+let hash = () =>
+  switch ([%external window]) {
+  | None => ""
+  | Some((window: Dom.window)) =>
+    switch (window |> location |> hash) {
+    | ""
+    | "#" => ""
+    | raw =>
+      /* remove the preceeding #, which every hash seems to have.
+         Why is this even included in location.hash?? */
+      raw |> Js.String.sliceToEnd(~from=1)
+    }
+  };
+let search = () =>
+  switch ([%external window]) {
+  | None => ""
+  | Some((window: Dom.window)) =>
+    switch (window |> location |> search) {
+    | ""
+    | "?" => ""
+    | raw =>
+      /* remove the preceeding ?, which every search seems to have. */
+      raw |> Js.String.sliceToEnd(~from=1)
+    }
+  };
+let push = path =>
+  switch ([%external history], [%external window]) {
+  | (None, _)
+  | (_, None) => ()
+  | (Some((history: Dom.history)), Some((window: Dom.window))) =>
+    pushState(history, ~href=path);
+    dispatchEvent(window, safeMakeEvent("popstate"));
+  };
+let replace = path =>
+  switch ([%external history], [%external window]) {
+  | (None, _)
+  | (_, None) => ()
+  | (Some((history: Dom.history)), Some((window: Dom.window))) =>
+    replaceState(history, ~href=path);
+    dispatchEvent(window, safeMakeEvent("popstate"));
+  };
+type url = {
+  path: list(string),
+  hash: string,
+  search: string,
+};
+type watcherID = unit => unit;
+let url = () => {path: path(), hash: hash(), search: search()};
+/* alias exposed publicly */
+let dangerouslyGetInitialUrl = url;
+let watchUrl = callback =>
+  switch ([%external window]) {
+  | None => (() => ())
+  | Some((window: Dom.window)) =>
+    let watcherID = () => callback(url());
+    addEventListener(window, "popstate", watcherID);
+    watcherID;
+  };
+let unwatchUrl = watcherID =>
+  switch ([%external window]) {
+  | None => ()
+  | Some((window: Dom.window)) =>
+    removeEventListener(window, "popstate", watcherID)
+  };
+
+let useUrl = () => {
+  let (url, setUrl) = React.useState(() => dangerouslyGetInitialUrl());
```

Comment:
we should be able to provide a default url here what do you think ? use case for example is if you are coming from ssr and want to start rendering with a different route.

Reply:
Wouldn't you do a redirect to the correct route?

Reply:
you mean `ReasonReact.Router.push('mynewroute');` instead ?

Reply:
No I mean on the server side issuing a 302 or whatever type of redirect it should be. It seems like you client and server should always be in agreement about how URLs map to routes and how that wraps to data.

Maybe I'm not understanding what you mean by "want to start rendering from a different route" - can you expand?

Reply:
want to avoid this: 
https://pennyworth.dblechoc.now.sh/more can you see the flashing ?

Reply:
express middleware
```
let app = <App client={client}  initialPath={here i can get in from the req knowing i am insdie an express middleware} />;
    // first render for ssr cache
    renderToStaticMarkup(app);
```

my actual component
```
[@react.component]
let make = (~client, ~inititalPath) => {

 // if i could pass an inititalPath here would be good
  let url = ReasonReactRouter.useUrl();

  <div>
    <GraphqlHooks.Provider value=client>
      <main>
        {switch (url.path |> mapPathToRoute) {
         | Home => <Home />
         | More => <More />
         }}
      </main>
    </GraphqlHooks.Provider>
  </div>;
};

```

Reply:
I see what you mean. Yeah I think this is reasonable. I think that `serverValue` or `staticValue` might be a better term - `initial` implies it's fixing an issue with the client not being able to pick up the URL when it first loaded, but it's actually an issue with getting the path via browser APIs on the server. But yes, adding a param here makes sense.

Reply:
awesome thank you

Reply:
https://github.com/reasonml/reason-react/pull/358